In [100]:
import sqlalchemy as sa
import pandas as pd

In [101]:
print(sa.__version__)

1.4.39


In [102]:
mmai_db_secret = {
    'drivername' : 'postgresql+psycopg2',
    'host'       : 'mmai5100postgres.canadacentral.cloudapp.azure.com',
    'port'       : '5432',
    'username'   : 'luwei',
    'password'   : '2023!Schulich',
    'database'   : 'mmai_db'
}

In [103]:
mmai_db_secret['username']

'luwei'

In [104]:
mmai_db_url = sa.engine.URL.create(
     drivername = mmai_db_secret['drivername'],
    username    = mmai_db_secret['username'],
    password    = mmai_db_secret['password'],
    host        = mmai_db_secret['host'],
    port        = mmai_db_secret['port'],
    database    = mmai_db_secret['database']
)

In [105]:
print(mmai_db_url)

postgresql+psycopg2://luwei:2023!Schulich@mmai5100postgres.canadacentral.cloudapp.azure.com:5432/mmai_db


In [106]:
mmai_db_engine = sa.create_engine(mmai_db_url)

In [107]:
with mmai_db_engine.connect() as connection:
    dimension_date_data = pd.read_sql(
        sql = 'SELECT * FROM dimensions.date_dimension;',
        con = connection
    )

In [108]:
dimension_date_data

sk_date        date   day_name  day_of_month  day_of_year  month  \
0     20190101  2019-01-01    Tuesday             1            1      1   
1     20190102  2019-01-02  Wednesday             2            2      1   
2     20190103  2019-01-03   Thursday             3            3      1   
3     20190104  2019-01-04     Friday             4            4      1   
4     20190105  2019-01-05   Saturday             5            5      1   
...        ...         ...        ...           ...          ...    ...   
1821  20231227  2023-12-27  Wednesday            27          361     12   
1822  20231228  2023-12-28   Thursday            28          362     12   
1823  20231229  2023-12-29     Friday            29          363     12   
1824  20231230  2023-12-30   Saturday            30          364     12   
1825  20231231  2023-12-31     Sunday            31          365     12   

     month_name  year year_week week  running_week year_quarter quarter  \
0       January  2019  2019-W01  W01             1      2019-Q1      Q1   
1       January  2019  2019-W01  W01             1      2019-Q1      Q1   
2       January  2019  2019-W01  W01             1      2019-Q1      Q1   
3       January  2019  2019-W01  W01             1      2019-Q1      Q1   
4       January  2019  2019-W01  W01             1      2019-Q1      Q1   
...         ...   ...       ...  ...           ...          ...     ...   
1821   December  2023  2023-W52  W52           261      2023-Q4      Q4   
1822   December  2023  2023-W52  W52           261      2023-Q4      Q4   
1823   December  2023  2023-W52  W52           261      2023-Q4      Q4   
1824   December  2023  2023-W52  W52           261      2023-Q4      Q4   
1825   December  2023  2023-W52  W52           261      2023-Q4      Q4   

      running_quarter  
0                   1  
1                   1  
2                   1  
3                   1  
4                   1  
...               ...  
1821               20  
1822               20  
1823               20  
1824               20  
1825               20  

[1826 rows x 14 columns]

In [128]:
dimension_date_data.dtypes

sk_date             int64
date               object
day_name           object
day_of_month        int64
day_of_year         int64
month               int64
month_name         object
year                int64
year_week          object
week               object
running_week        int64
year_quarter       object
quarter            object
running_quarter     int64
dtype: object

In [109]:
import sqlalchemy as sa
import pandas as pd

In [110]:
my_db_secret = {
    'drivername' : 'postgresql+psycopg2',
    'host'       : 'mmai5100postgres.canadacentral.cloudapp.azure.com',
    'port'       : '5432',
    'username'   : 'luwei',
    'password'   : '2023!Schulich',
    'database'   : 'luwei_db'
}

In [111]:
my_db_url = sa.engine.URL.create(
     drivername = my_db_secret['drivername'],
    username    = my_db_secret['username'],
    password    = my_db_secret['password'],
    host        = my_db_secret['host'],
    port        = my_db_secret['port'],
    database    = my_db_secret['database']
)

In [112]:
my_db_engine = sa.create_engine(my_db_url)

In [113]:
date_column = dimension_date_data['sk_date']

In [114]:
date_column

0       20190101
1       20190102
2       20190103
3       20190104
4       20190105
          ...   
1821    20231227
1822    20231228
1823    20231229
1824    20231230
1825    20231231
Name: sk_date, Length: 1826, dtype: int64

In [115]:
with my_db_engine.connect() as connection:
    connection.execute('CREATE SCHEMA IF NOT EXISTS nervous;')

In [116]:
dimension_date_data.to_sql(
    name = 'dimension_date_data',
    con = my_db_engine,
    schema = 'nervous',
    if_exists = 'replace',
    index = False,
    method = 'multi',
 dtype = {
    'sk_date': sa.types.VARCHAR(25),
    'date': sa.types.DATE, 
    'day_name': sa.types.VARCHAR(25),
    'day_of_month': sa.types.VARCHAR(25),
    'day_of_year': sa.types.VARCHAR(25),
    'month_name':sa.types.VARCHAR(25),
    'month': sa.types.VARCHAR(25),
    'year': sa.types.VARCHAR(25),
    'year_week': sa.types.VARCHAR(25),
    'week': sa.types.VARCHAR(25),
    'running_week': sa.types.VARCHAR(25),
    'year_quarter': sa.types.VARCHAR(25),
    'quarter': sa.types.VARCHAR(25),
    'running_quarter': sa.types.VARCHAR(25),
}
)

1826

0190101: Integer or String (depending on usage; if you plan to perform date operations, storing it as a string and then converting to a date format might be more useful).
2019-01-01: Date (in databases, this would be a DATE data type; in programming languages or data analysis tools like pandas, you'd use a date or datetime type).
Tuesday: String (to represent the day of the week).
1: Integer (assuming this represents a day, week, month, quarter, or some other countable unit).
1: Integer (similar reasoning as above).
1: Integer (similar reasoning as above).
January: String (to represent the month name).
2019: Integer (to represent the year).
2019-W01: String (to represent a week in a specific format).
W01: String (to represent the week number).
1: Integer (similar reasoning as previous integer values; this might represent a week, month, or quarter number).
2019-Q1: String (to represent a quarter in a specific format).
Q1: String (to represent the quarter number).
1: Integer (similar reasoning as previous integer values).
Remember that the exact data type can depend on the context and the operations you want to perform on the data. For instance, if you're not performing arithmetic on the "20190101" value and you're instead using it as a unique identifier, you might prefer to store it as a string. Similarly, if you're working within a specific database or programming language, you'd choose the data type that best aligns with your intended operations and storage constraints.


In [117]:
import sqlalchemy as sa
import pandas as pd
import requests, json

In [119]:
resource_id = '8a89caa9-511c-4568-af89-7f2174b4378c'
limit = 500

api_url = 'https://data.ontario.ca/api/3/action/datastore_search?resource_id={}&limit={}'.format(resource_id,limit)
print(api_url)


https://data.ontario.ca/api/3/action/datastore_search?resource_id=8a89caa9-511c-4568-af89-7f2174b4378c&limit=500


In [120]:
data = requests.get(api_url)



In [121]:
data

<Response [200]>

In [122]:
data = data.json()
data

{'help': 'https://data.ontario.ca/api/3/action/help_show?name=datastore_search',
 'success': True,
 'result': {'include_total': True,
  'limit': 500,
  'records_format': 'objects',
  'resource_id': '8a89caa9-511c-4568-af89-7f2174b4378c',
  'total_estimation_threshold': None,
  'records': [{'_id': 1,
    'report_date': '2021-01-01T00:00:00',
    'previous_day_total_doses_administered': 6252,
    'previous_day_at_least_one': 6242,
    'previous_day_fully_vaccinated': 2,
    'previous_day_3doses': None,
    'total_doses_administered': 32210,
    'total_individuals_at_least_one': '32155',
    'total_individuals_partially_vaccinated': '32144',
    'total_doses_in_fully_vaccinated_individuals': '22',
    'total_individuals_fully_vaccinated': '11',
    'total_individuals_3doses': None},
   {'_id': 2,
    'report_date': '2021-01-02T00:00:00',
    'previous_day_total_doses_administered': 4615,
    'previous_day_at_least_one': 4603,
    'previous_day_fully_vaccinated': 6,
    'previous_day_3dose

In [123]:
data['help']

'https://data.ontario.ca/api/3/action/help_show?name=datastore_search'

In [124]:
data['success']

True

In [127]:
data['result']

{'include_total': True,
 'limit': 500,
 'records_format': 'objects',
 'resource_id': '8a89caa9-511c-4568-af89-7f2174b4378c',
 'total_estimation_threshold': None,
 'records': [{'_id': 1,
   'report_date': '2021-01-01T00:00:00',
   'previous_day_total_doses_administered': 6252,
   'previous_day_at_least_one': 6242,
   'previous_day_fully_vaccinated': 2,
   'previous_day_3doses': None,
   'total_doses_administered': 32210,
   'total_individuals_at_least_one': '32155',
   'total_individuals_partially_vaccinated': '32144',
   'total_doses_in_fully_vaccinated_individuals': '22',
   'total_individuals_fully_vaccinated': '11',
   'total_individuals_3doses': None},
  {'_id': 2,
   'report_date': '2021-01-02T00:00:00',
   'previous_day_total_doses_administered': 4615,
   'previous_day_at_least_one': 4603,
   'previous_day_fully_vaccinated': 6,
   'previous_day_3doses': None,
   'total_doses_administered': 36825,
   'total_individuals_at_least_one': '36706',
   'total_individuals_partially_vaccin

In [135]:
covid_dose = data['result']['records']

In [136]:
covid_dose = pd.DataFrame(covid_dose)
covid_dose

_id          report_date  previous_day_total_doses_administered  \
0      1  2021-01-01T00:00:00                                   6252   
1      2  2021-01-02T00:00:00                                   4615   
2      3  2021-01-03T00:00:00                                   5396   
3      4  2021-01-04T00:00:00                                   5996   
4      5  2021-01-05T00:00:00                                   9687   
..   ...                  ...                                    ...   
495  496  2022-05-11T00:00:00                                  28955   
496  497  2022-05-12T00:00:00                                  28244   
497  498  2022-05-13T00:00:00                                  27598   
498  499  2022-05-14T00:00:00                                  25694   
499  500  2022-05-15T00:00:00                                  14851   

     previous_day_at_least_one  previous_day_fully_vaccinated  \
0                         6242                              2   
1                         4603                              6   
2                         5389                              3   
3                         5993                              0   
4                         9662                             21   
..                         ...                            ...   
495                        837                           1087   
496                        799                           1086   
497                        878                           1119   
498                        807                           1100   
499                        751                           1276   

    previous_day_3doses  total_doses_administered  \
0                  None                     32210   
1                  None                     36825   
2                  None                     42221   
3                  None                     48217   
4                  None                     57904   
..                  ...                       ...   
495                2656                  33164767   
496                2515                  33193011   
497                2469                  33220609   
498                2610                  33246303   
499                2058                  33261154   

    total_individuals_at_least_one total_individuals_partially_vaccinated  \
0                            32155                                  32144   
1                            36706                                  36693   
2                            42095                                  42079   
3                            48087                                  48072   
4                            57747                                  57712   
..                             ...                                    ...   
495                       12593790                                 451943   
496                       12594292                                 451652   
497                       12594859                                 451406   
498                       12595352                                 451106   
499                       12595831                                 450574   

    total_doses_in_fully_vaccinated_individuals  \
0                                            22   
1                                            26   
2                                            32   
3                                            30   
4                                            70   
..                                          ...   
495                                        None   
496                                        None   
497                                        None   
498                                        None   
499                                        None   

    total_individuals_fully_vaccinated total_individuals_3doses  
0                                   11                     None  
1                                   13              

In [141]:
new_doses = covid_dose['total_doses_administered'] - covid_dose['previous_day_total_doses_administered']
new_doses

0         25958
1         32210
2         36825
3         42221
4         48217
         ...   
495    33135812
496    33164767
497    33193011
498    33220609
499    33246303
Length: 500, dtype: int64

In [143]:
covid_dose['previous_day_at_least_one']

0      6242
1      4603
2      5389
3      5993
4      9662
       ... 
495     837
496     799
497     878
498     807
499     751
Name: previous_day_at_least_one, Length: 500, dtype: int64

In [144]:
covid_dose['total_individuals_at_least_one']

0         32155
1         36706
2         42095
3         48087
4         57747
         ...   
495    12593790
496    12594292
497    12594859
498    12595352
499    12595831
Name: total_individuals_at_least_one, Length: 500, dtype: object

In [147]:
def to_integer(val):
    try:
        return int(val)
    except ValueError:
        return -1  # or any default integer value


In [148]:
covid_dose['total_individuals_at_least_one'] = covid_dose['total_individuals_at_least_one'].apply(to_integer)

In [149]:
covid_dose['total_individuals_at_least_one']

0         32155
1         36706
2         42095
3         48087
4         57747
         ...   
495    12593790
496    12594292
497    12594859
498    12595352
499    12595831
Name: total_individuals_at_least_one, Length: 500, dtype: int64

In [151]:
new_doses_unvaccinated = covid_dose['total_individuals_at_least_one'] - covid_dose['previous_day_at_least_one']
new_doses_unvaccinated

0         25913
1         32103
2         36706
3         42094
4         48085
         ...   
495    12592953
496    12593493
497    12593981
498    12594545
499    12595080
Length: 500, dtype: int64